flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java

```diff
@@ -288,10 +288,24 @@ private void configureStreamGraph(final StreamGraph graph) {
 		} else {
 			graph.setStateBackend(stateBackend);
 			graph.setGlobalDataExchangeMode(GlobalDataExchangeMode.ALL_EDGES_PIPELINED);
+			graph.setAllVerticesInSameSlotSharingGroupByDefault(true);
 			graph.setScheduleMode(ScheduleMode.EAGER);
+
+			if (checkpointConfig.isApproximateLocalRecoveryEnabled()) {
+				checkApproximateLocalRecoveryCompatibility();
+				graph.setGlobalDataExchangeMode(GlobalDataExchangeMode.ALL_EDGES_PIPELINED_APPROXIMATE);
+			} else {
+				graph.setGlobalDataExchangeMode(GlobalDataExchangeMode.ALL_EDGES_PIPELINED);
+			}
 		}
 	}
 
+	private void checkApproximateLocalRecoveryCompatibility() {
+		checkState(
+			!checkpointConfig.isUnalignedCheckpointsEnabled(),
+			"Approximate Local Recovery and Unaligned Checkpoint can not be used together yet");
```

Comment:
Can we do a similar check for the scheduler configuration? I think we should fail hard if we see that we are not using the `legacy` scheduler. Moreover, an explanatory exception message would be helpful as well.

Reply:
I've tried out different ways to do the check. It is slightly different from the unaligned checkpoint check, because the PipelinedRegionScheduler is configured in the JobMaster level, and approximate local recovery is configured at the job level.

The check should happen in the same place where either PIPELINED_REGION_SCHEDULING or LEGACY_SCHEDULING is decided, where the type SchedulingStrategyFactory is decided.

1. In JobGraph, the most reasonable place to put the config "isApproximateLocalRecoveryEnabled" seems to be `JobCheckpointingSettings#CheckpointCoordinatorConfiguration`, similar to unaligned checkpoint's config. However `CheckpointCoordinatorConfiguration` as its name, is for CheckpointCoordinator and will be serialized to CheckpointCoordinator. But in fact, CheckpointCoordinator does not need `isApproximateLocalRecoveryEnabled` for anything, and it breaks a lot of tests, so, at this point, it is probably the best place to put.

2. So I put `isApproximateLocalRecoveryEnabled` in a similar place as `scheduleMode` in JobGraph, and will be removed together with `scheduleMode` later. This flag is only used to make sure ApproximateLocalRecovery is not used together with JobManagerOptions.SCHEDULING_STRATEGY to region

3. If JobManagerOptions.SCHEDULING_STRATEGY is set to legacy, EAGER strategy is enforced in `StreamGraphGenerator#configureStreamGraph`

```
} else {
			graph.setStateBackend(stateBackend);
			graph.setScheduleMode(ScheduleMode.EAGER);

			if (checkpointConfig.isApproximateLocalRecoveryEnabled()) {
				checkApproximateLocalRecoveryCompatibility();
				graph.setGlobalDataExchangeMode(GlobalDataExchangeMode.ALL_EDGES_PIPELINED_APPROXIMATE);
			} else {
				graph.setGlobalDataExchangeMode(GlobalDataExchangeMode.ALL_EDGES_PIPELINED);
			}
		}
```

Reply:
I will also add a test for this.

Reply:
Yes I think that we need to do the check on the cluster/JobMaster-side. At the moment we haven't exposed this configuration on a per-job basis because ideally we only have a single scheduler which is able to serve all workloads.